Add unit tests

nielslange opened the pull request:
Fixes #2 

### Notes

Currently, the components are not covered by unit tests. The only way to find out if a future development breaks existing functionality, is to manually test the functionality.

This PR adds unit tests for the `<ProgressBar>` and `<ProgressLabel>` components.

### Steps to test

1. Check out this PR.
2. Run `nmp ci`.
3. Run `npm test`.
4. Verify that unit tests have successfully been executed.

### Additional notes

Besides the command `npm test`, I also introduced the following commands:
- `npm run test:coverage` → Run tests and show coverage report.
- `npm run test:help` → Show help for the unit tests.
- `npm run test:watch` → Run tests and watch for changes.

nielslange committed:
Add unit tests

senadir commented on src/components/progress-bar/__test__/index.test.js:
```diff
@@ -0,0 +1,69 @@
+describe( 'The ProgressBar component', () => {
+	it( 'renders without crashing', () => {
+		const div = document.createElement( 'div' );
+		ReactDOM.render( <ProgressBar></ProgressBar>, div );
```
Shouldn't you be using `render` here?
Also this test has no expect, I know you're not expecting anything to happen, which is why you should do an `expect.not.toThrow`, using `render` from React testing library.

nielslange replied:
I added this test based on https://create-react-app.dev/docs/running-tests/#testing-components. That said, that section also mentioned the following:

> If you haven’t decided on a testing strategy yet, we recommend that you start with creating basic smoke tests for your components:

```
import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';

it('renders without crashing', () => {
  const div = document.createElement('div');
  ReactDOM.render(<App />, div);
});
```

Given that I have other tests that are calling the `<ProgressBar />` component, I went ahead and removed the `renders without crashing` test cases from both test suits.

senadir replied:
It's true that this test is redundant, my question was mostly around using `document.createElement` and `ReactDOM.render` instead of react-testing-library render function.

senadir commented on src/components/progress-bar/__test__/index.test.js:
```diff
@@ -0,0 +1,69 @@
+	} );
+
+	it( 'shows the classnames of the outer div correctly', () => {
+		const { getByTestId } = render( <ProgressBar></ProgressBar> );
+		expect( getByTestId( 'progress-bar-outer' ) ).toHaveClass(
+			'wc-free-shipping-progress-bar__outer'
+		);
+	} );
```
Using a testId here is not needed, you can just return the `container` from the render call, and have it in the expect check.

senadir replied:
Also, this runs by the assumption that the class name won't change, is this what you're trying to test here?

nielslange replied:
Thanks for pointing this out. I replaced `getByTestId()` with `document.querySelector()` to keep the code lean.

senadir commented on src/components/progress-bar/__test__/index.test.js:
```diff
@@ -0,0 +1,69 @@
+	} );
+
+	it( 'shows the classnames of the inner div correctly', () => {
+		const { getByTestId } = render( <ProgressBar></ProgressBar> );
+		expect( getByTestId( 'progress-bar-inner' ) ).toHaveClass(
+			'wc-free-shipping-progress-bar__inner'
+		);
+	} );
```
As above, you can use `container.querySelector( '.wc-free-shipping-progress-bar__inner' ).toBeInTheDocument();

nielslange replied:
As before, I replaced `getByTestId()` for all test cases.

senadir commented on src/components/progress-bar/__test__/index.test.js:
```diff
@@ -0,0 +1,69 @@
+		const { getByTestId } = render(
+			<ProgressBar freeShippingFrom="4" currentTotal="4"></ProgressBar>
+		);
+		expect( getByTestId( 'progress-bar-inner' ) ).toHaveStyle(
```
From React testing library:
> it is recommended to use [test id] only after the other queries don't work for your use case. Using data-testid attributes do not resemble how your software is used and should be avoided if possible.

Which is why I would recommend trying with something stable but isn't a test id.

nielslange replied:
Given that I'm able to target all elements via CSS classes, I removed the `data-testid` attributes completely.

senadir commented on src/components/progress-label/__test__/index.test.js:
```diff
@@ -0,0 +1,52 @@
+				currentTotal="1"
+			></ProgressLabel>
+		);
+		expect( getByTestId( 'progress-bar-label' ) ).toHaveTextContent(
```
You can just query be test here "getByText"

nielslange replied:
Sweet! This makes this test case way more readable!

Before:

```js
it( 'shows the message for insufficient totals correctly', () => {
	const { getByTestId } = render(
		<ProgressLabel
			labelInsufficientTotals="Spend only {amount} more to get free shipping!"
			freeShippingFrom="2"
			currentTotal="1"
		></ProgressLabel>
	);
	expect( getByTestId( 'progress-bar-label' ) ).toHaveTextContent(
		'Spend only 1.00 more to get free shipping!'
	);
} );
```

After:

```js
it( 'shows the message for insufficient totals correctly', () => {
	render(
		<ProgressLabel
			labelInsufficientTotals="Spend only {amount} more to get free shipping!"
			freeShippingFrom="2"
			currentTotal="1"
		/>
	);
	screen.getByText( 'Spend only 1.00 more to get free shipping!' );
} );

```

senadir reviewed (changes requested):
I left a comment on both of tests, you should avoid using test id unless you must, I also left a couple of comments about how to render and other type of queries.

nielslange committed:
Remove data-testid

nielslange committed:
Remove obsolete test case and replace getByTestId() with querySelector()

nielslange committed:
Remove obsolete dependencies

nielslange committed:
Replace toHaveTextContent() with getByText()

nielslange commented:
> I left a comment on both of tests, you should avoid using test id unless you must, I also left a couple of comments about how to render and other type of queries.

Thanks for your quick review. I addressed all points and pushed the changes. Would you mind reviewing the changes again, @senadir?

senadir commented on src/components/progress-bar/__test__/index.test.js:
```diff
@@ -0,0 +1,45 @@
+ */
+import ProgressBar from './../index';
+
+const outer = '.wc-free-shipping-progress-bar__outer';
+const inner = '.wc-free-shipping-progress-bar__inner';
+
+describe( 'The ProgressBar component', () => {
+	it( 'shows the classnames of the outer div correctly', () => {
+		render( <ProgressBar /> );
+		expect( document.querySelector( outer ) ).toBeInTheDocument();
+	} );
+
+	it( 'shows the classnames of the inner div correctly', () => {
+		render( <ProgressBar /> );
+		expect( document.querySelector( inner ) ).toBeInTheDocument();
+	} );
```
This is not a blocker, just a general question/caveat.

We're not testing implementation details here, unless `.wc-free-shipping-progress-bar__outer` is part of your public API. I suppose given this is div, there isn't much we can to select it by, which I'm okay with. However, in future projects, consider other ways to test this. If this test if purely visual, maybe an image regression test would work better?

nielslange replied:
> We're not testing implementation details here, unless .wc-free-shipping-progress-bar__outer is part of your public API. I suppose given this is div, there isn't much we can to select it by, which I'm okay with. However, in future projects, consider other ways to test this. If this test if purely visual, maybe an image regression test would work better?

Both `.wc-free-shipping-progress-bar__outer` and `.wc-free-shipping-progress-bar__inner` are needed to display the progress bar. The outer div defines the absolute height of the progress bar, is needed to show the progress in percentage. If one of these two CSS classes isn't loading or gets renamed, the progress bar will no longer be working correctly, That's why I added checks for the CSS classes in this case.

Thanks for pointing out the option of replacing them with visual regression tests. I looked up visual regression tests briefly, and they seem to require Puppeteer. As I don't have e2e tests in place at the moment, I'd go ahead and merge this PR as it is. But I'll add a note to the e2e issue to potentially replace the CSS class checks of this PR with a visual regression test.

senadir reviewed (approved):
This is looking good! thank you for updating the tests Niels.
